File: leetcode/114.flatten_binary_tree_to_linked_list.rb
'''
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
'''
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def flatten(root)
  return root if root.nil?
  dummy = root
  right = dummy.right
  dummy.right = flatten(dummy.left)
  dummy.left = nil
  dummy = dummy.right until dummy.right.nil?
  dummy.right = flatten(right)
  root
end

tree = TreeNode.new(1)
tree.left = TreeNode.new(2)
tree.right = TreeNode.new(5)
left = tree.left
right = tree.right
left.left = TreeNode.new(3)
left.right = TreeNode.new(4)
right.right = TreeNode.new(6)

p flatten(tree)

File: leetcode/pythonSolutions/114.flatten_binary_tree_to_linked_list.py
'''
114. Flatten Binary Tree to Linked List
Medium

Given a binary tree, flatten it to a linked list in-place.
For example, given the following tree:

    1
   / \
  2   5
 / \   \
3   4   6
The flattened tree should look like:

1
 \
  2
   \
    3
     \
      4
       \
        5
         \
          6
'''

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        oldRight = root.right
        root.right = root.left        
        self.flatten(root.left) 
        root.left = None
        while root.right:
            root = root.right
        root.right = oldRight
        self.flatten(oldRight)

def prt(node):
  if node:
    print(node.val)
    prt(node.left)
    prt(node.right)

node1 = TreeNode(1)
node2 = TreeNode(2)
node5 = TreeNode(5)
node1.left = node2
node1.right = node5


sol = Solution()
sol.flatten(node1)
prt(node1)